### ISSUE TITLE
calwebb_spec2 background step failing for NIS_WFSS exposure

### ISSUE THREAD

**Issue by stscijgbot-jp** [spacetelescope/jwst#4932]
_Issue [JP-1447](https://jira.stsci.edu/browse/JP-1447) was created on JIRA by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse):_

The calwebb_spec2 pipeline is failing in the background step for a NIS_WFSS exposure when it tries to load the source catalog (to find background regions in the image) with the following traceback:
```java
2020-05-12 02:16:43,132 - stpipe.Spec2Pipeline.bkg_subtract - INFO - Extracting objects < abmag = 99.0
2020-05-12 02:16:43,132 - stpipe.Spec2Pipeline.bkg_subtract - INFO - Getting objects from /ifs/archive/test/jwst/store/doggett/tests/complete_long_05/dtest2/JWSTDP-2020.2.0-51~dc4bb2fc/2020-05-11-185809/sdp/asn_creation/cal/level2/jw00625-o025_t001_niriss_clear-f150w_cat.ecsv
2020-05-12 02:16:43,253 - stpipe.Spec2Pipeline.bkg_subtract - ERROR - Missing required columns in source catalog: {'nn_dist', 'aper70_abmag', 'aper50_vegamag', 'aper30_abmag_err', 'aper50_abmag', 'CI_30_70', 'aper_total_flux', 'isophotal_vegamag', 'sharpness', 'isophotal_vegamag_err', 'sky_orientation', 'aper50_abmag_err', 'is_star', 'aper_total_abmag_err', 'isophotal_flux', 'aper70_vegamag', 'aper30_flux_err', 'aper30_vegamag', 'aper_bkg_flux_err', 'aper_total_vegamag', 'semimajor_sigma', 'aper70_flux_err', 'aper50_vegamag_err', 'isophotal_area', 'aper_total_flux_err', 'ellipticity', 'aper70_flux', 'aper50_flux_err', 'aper30_vegamag_err', 'semiminor_sigma', 'aper70_vegamag_err', 'aper_bkg_flux', 'isophotal_abmag', 'aper_total_vegamag_err', 'isophotal_flux_err', 'aper30_abmag', 'aper50_flux', 'aper30_flux', 'aper_total_abmag', 'orientation', 'nn_abmag', 'isophotal_abmag_err', 'aper70_abmag_err', 'roundness'}
Traceback (most recent call last):
  File "/dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-0.16.0.20200507-py37/lib/python3.7/site-packages/jwst/pipeline/calwebb_spec2.py", line 95, in process
    asn.filename
  File "/dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-0.16.0.20200507-py37/lib/python3.7/site-packages/jwst/pipeline/calwebb_spec2.py", line 217, in process_exposure_product
    input = self.bkg_subtract(input, bkg_list)
  File "/dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-0.16.0.20200507-py37/lib/python3.7/site-packages/jwst/stpipe/step.py", line 392, in run
    step_result = self.process(*args)
  File "/dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-0.16.0.20200507-py37/lib/python3.7/site-packages/jwst/background/background_step.py", line 58, in process
    input_model, bkg_name, wlrange_name)
  File "/dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-0.16.0.20200507-py37/lib/python3.7/site-packages/jwst/background/background_sub.py", line 145, in subtract_wfss_bkg
    bkg_mask = mask_from_source_cat(input_model, wl_range_name)
  File "/dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-0.16.0.20200507-py37/lib/python3.7/site-packages/jwst/background/background_sub.py", line 229, in mask_from_source_cat
    grism_obj_list = create_grism_bbox(input_model, reference_files)
  File "/dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-0.16.0.20200507-py37/lib/python3.7/site-packages/jwst/assign_wcs/util.py", line 560, in create_grism_bbox
    skyobject_list = get_object_info(input_model.meta.source_catalog)
  File "/dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-0.16.0.20200507-py37/lib/python3.7/site-packages/jwst/assign_wcs/util.py", line 459, in get_object_info
    raise KeyError(err_text)
KeyError: "Missing required columns in source catalog: {'nn_dist', 'aper70_abmag', 'aper50_vegamag', 'aper30_abmag_err', 'aper50_abmag', 'CI_30_70', 'aper_total_flux', 'isophotal_vegamag', 'sharpness', 'isophotal_vegamag_err', 'sky_orientation', 'aper50_abmag_err', 'is_star', 'aper_total_abmag_err', 'isophotal_flux', 'aper70_vegamag', 'aper30_flux_err', 'aper30_vegamag', 'aper_bkg_flux_err', 'aper_total_vegamag', 'semimajor_sigma', 'aper70_flux_err', 'aper50_vegamag_err', 'isophotal_area', 'aper_total_flux_err', 'ellipticity', 'aper70_flux', 'aper50_flux_err', 'aper30_vegamag_err', 'semiminor_sigma', 'aper70_vegamag_err', 'aper_bkg_flux', 'isophotal_abmag', 'aper_total_vegamag_err', 'isophotal_flux_err', 'aper30_abmag', 'aper50_flux', 'aper30_flux', 'aper_total_abmag', 'orientation', 'nn_abmag', 'isophotal_abmag_err', 'aper70_abmag_err', 'roundness'}"
Traceback (most recent call last):
  File "/dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-0.16.0.20200507-py37/bin/strun", line 26, in <module>
    step = Step.from_cmdline(sys.argv[1:])
  File "/dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-0.16.0.20200507-py37/lib/python3.7/site-packages/jwst/stpipe/step.py", line 156, in from_cmdline
    return cmdline.step_from_cmdline(args)
  File "/dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-0.16.0.20200507-py37/lib/python3.7/site-packages/jwst/stpipe/cmdline.py", line 346, in step_from_cmdline
    step.run(*positional)
  File "/dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-0.16.0.20200507-py37/lib/python3.7/site-packages/jwst/stpipe/step.py", line 392, in run
    step_result = self.process(*args)
  File "/dms/local/jwst/pipeline/pkgs/miniconda3/envs/jwstdp-0.16.0.20200507-py37/lib/python3.7/site-packages/jwst/pipeline/calwebb_spec2.py", line 111, in process
    'One or more products failed to process. Failing calibration.'
RuntimeError: One or more products failed to process. Failing calibration.
----------------------------------------------------------------------
ERROR RUNNING STEP 'Spec2Pipeline':
    One or more products failed to process. Failing calibration.
----------------------------------------------------------------------
```
suggesting that the course catalog is missing some expected columns of data.

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-1447?focusedCommentId=463488#comment-463488):_

The input spec2 ASN file, rate product for the grism image, and the source cat file have been copied to the data directory listed above. A quick view of the source cat file suggests that the columns claimed to be missing appear to be there.

**Comment by stscijgbot-jp** [comment]
_Comment by [James White](https://jira.stsci.edu/secure/ViewProfile.jspa?name=jwhite) on [JIRA](https://jira.stsci.edu/browse/JP-1447?focusedCommentId=463815#comment-463815):_

The set difference was just in the wrong order. Turns out the missing keys are actually 'abmag_error' and 'abmag' which are both "required" fields

**Comment by stscijgbot-jp** [comment]
_Comment by [James White](https://jira.stsci.edu/secure/ViewProfile.jspa?name=jwhite) on [JIRA](https://jira.stsci.edu/browse/JP-1447?focusedCommentId=463817#comment-463817):_

So, with the data provided, it will still fail until those columns are available in the given source catalogue.

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-1447?focusedCommentId=463837#comment-463837):_

Huh, but those should be two of the fundamental columns always included in a source catalog.

[Larry Bradley](https://jira.stsci.edu/secure/ViewProfile.jspa?name=lbradley) there's code in other parts of the pipeline that is looking for columns "abmag" and "abmag_error" in the catalogs created by the new source_catalog step. Do those columns no longer exist? It looks like there's now multiple columns called "aper30_abmag", "aper50_abmag", "aper70_abmag", and "aper_total_abmag" (each with corresponding _err columns). Should that code be updated to now use "aper_total_abmag" and "aper_total_abmag_err"? Are those semi-equivalent to the old "abmag" and "abmag_error" columns?

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-1447?focusedCommentId=463838#comment-463838):_

The code in question is here: <https://github.com/spacetelescope/jwst/blob/master/jwst/lib/catalog_utils.py#L78>

Can you tell us which of those columns no longer exist and/or have had their names changed?

**Comment by stscijgbot-jp** [comment]
_Comment by [Larry Bradley](https://jira.stsci.edu/secure/ViewProfile.jspa?name=lbradley) on [JIRA](https://jira.stsci.edu/browse/JP-1447?focusedCommentId=463843#comment-463843):_

The old "abmag" and "abmag_error" columns translate to the new "isophotal_abmag" and "isophotal_abmag_err" columns.

The other columns listed in <https://github.com/spacetelescope/jwst/blob/master/jwst/lib/catalog_utils.py#L78> are unchanged, except there never was an "sid" column.  It's always been "id", so perhaps somewhere else the code translates from "id" to "sid".

For reference the "aperXX_abmag" columns represent the circular aperture photometry for the XX encircled energy.  And "aper_total" is the aperture-corrected total circular aperture photometry.

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-1447?focusedCommentId=463846#comment-463846):_

Great - thanks! (y)

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-1447?focusedCommentId=463860#comment-463860):_

The "sid" to "id" translation appears to take place here:

<https://github.com/spacetelescope/jwst/blob/master/jwst/assign_wcs/util.py#L451>

so that one should be OK.

**Comment by stscijgbot-jp** [comment]
_Comment by [James White](https://jira.stsci.edu/secure/ViewProfile.jspa?name=jwhite) on [JIRA](https://jira.stsci.edu/browse/JP-1447?focusedCommentId=463874#comment-463874):_

I've updated the old keys and the Spec2Pipeline now runs to completion on the data referenced above.

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-1447?focusedCommentId=463962#comment-463962):_

Fixed by #4943
